allensdk/brain_observatory/filtered_running_speed/filtered_running_speed.py

```diff
@@ -0,0 +1,382 @@
+        self
+    ):
+        """
+        Process an experiment with a single simulus sessions
+        """
+
+        start_index = FilteredRunningSpeed.START_FRAME
+
+        frame_times = self._get_frame_times()
+
+        end_index = len(frame_times)
+
+        velocities, raw_data = self._extract_dx_info(
+            frame_times,
+            start_index,
+            end_index,
+            self.pkl_path
+        )
+
+        store = pd.HDFStore(self.args['output_path'])
+        store.put("running_speed", velocities)
+        store.put("raw_data", raw_data)
+        store.close()
+
+        self._write_output_file()
+
+    def _write_output_file(self):
+        """
+        Write the output json file
+        """
+
+        ouput_data = {}
+        ouput_data['output_path'] = self.args['output_path']
+        ouput_data['input_parameters'] = self.args
+
+        with open(self.args['output_json'], 'w') as output_file:
+            json.dump(ouput_data, output_file, indent=2)
+
+    def _process_multi_simulus_experiment(
+        self
+    ):
+        """
+        Process an experiment with a three simulus sessions
+        """
+
+        (
+            behavior_start,
+            behavior_end,
+            mapping_start,
+            mapping_end,
+            replay_start,
+            replay_end
+        ) = self._get_stimulus_starts_and_ends()
+
+        frame_times = self._get_frame_times()
+
+        behavior_velocities = self._extract_dx_info(
+            frame_times,
+            behavior_start,
+            behavior_end,
+            self.args['behavior_pkl_path']
+        )
+
+        mapping_velocities = self._extract_dx_info(
+            frame_times,
+            mapping_start,
+            mapping_end,
+            self.args['mapping_pkl_path']
+        )
+
+        replay_velocities = self._extract_dx_info(
+            frame_times,
+            replay_start,
+            replay_end,
+            self.args['replay_pkl_path']
+        )
+
+        velocities, raw_data = self._merge_dx_data(
+            mapping_velocities,
+            behavior_velocities,
+            replay_velocities,
+            frame_times
+        )
+
+        store = pd.HDFStore(self.args['output_path'])
+        store.put("running_speed", velocities)
+        store.put("raw_data", raw_data)
+        store.close()
+
+        self._write_output_file()
+
+    def process(self, default_zscore_threshold, use_lowpass_filter):
+        """
+        Process an experiment
+        """
+
+        self.default_zscore_threshold = default_zscore_threshold
+        self.use_lowpass_filter = use_lowpass_filter
+
+        if self.args['mapping_pkl_path'] is not None:
+            if (self.args['behavior_pkl_path'] is not None
+                    and self.args['replay_pkl_path'] is not None):
+                self._process_multi_simulus_experiment()
+            else:
+                self._process_single_simulus_experiment()
```

Comment:
How does the single-stimulus case differ from the implementation in `brain_observatory/behavior/data_objects/running_speed.py`? Is the only difference the ability to clip the timesteps to `start_index:end_index`. If that's the only difference, it seems like it would be worth adding that option to the existing implementation, since the `to_nwb` functionality will be ready for us in that case.

This assumes that you want the same dataframe that the other implementation constructs, but it looks like you do.

Reply:
In that vein, it seems like the most self-consistent design choice is to implement a `MultiStimulusRunningSpeed` class that mimics the existing `RunningSpeed` class, and then have the `FilteredRunningSpeed` CLI module instantiate either `MultiStimulusRunningSpeed` or `RunningSpeed`, depending on the input.

That is probably another half week of work, and I don't know how urgently we want this pipeline to go live

@wbwakeman ?

Reply:
I've removed the single-stimulus case. My code now only supports the multi stimulus case.

Reply:
as commented on by @wbwakeman in the stand up - I'm going to create an issue for the this refactor and then hold off on it until later.